File: sys/yellowstone/hw/gpt.cpp
#include "hw/gpt.h"

#include <glacier/memory/move.h>
#include <glacier/status/error.h>
#include <mammoth/debug.h>
#include <mammoth/memory_region.h>
#include <zcall.h>
#include <zglobal.h>

#define GPT_DEBUG 0

const uint64_t kSectorSize = 512;

const uint64_t kGptPartitionSignature = 0x54524150'20494645;

const uint64_t kLfsDataLow = 0x477284830fc63daf;
const uint64_t kLfsDataHigh = 0xe47d47d8693d798e;

struct MbrPartition {
  uint8_t boot_indicator;
  uint8_t starting_chs[3];
  uint8_t os_type;
  uint8_t ending_chs[3];
  uint32_t starting_lba;
  uint32_t ending_lba;
} __attribute__((packed));

struct ParititionHeader {
  uint64_t signature;
  uint32_t revision;
  uint32_t header_size;
  uint32_t crc_32;
  uint32_t reserved;
  uint64_t lba_self;
  uint64_t lba_mirror;
  uint64_t lba_min;
  uint64_t lba_max;
  uint64_t guid_low;
  uint64_t guid_high;
  uint64_t lba_partition_entries;
  uint32_t num_partitions;
  uint32_t parition_entry_size;
  uint32_t partition_entry_crc32;
} __attribute__((packed));

struct PartitionEntry {
  uint64_t type_guid_low;
  uint64_t type_guid_high;
  uint64_t part_guid_low;
  uint64_t part_guid_high;
  uint64_t lba_start;
  uint64_t lba_end;
  uint64_t attributes;
  char partition_name[72];
} __attribute__((packed));

GptReader::GptReader(glcr::UniquePtr<DenaliClient> denali)
    : denali_(glcr::Move(denali)) {}

glcr::ErrorCode GptReader::ParsePartitionTables() {
  ReadRequest req;
  req.set_device_id(0);
  req.set_lba(0);
  req.set_size(2);
  ReadResponse resp;
  RET_ERR(denali_->Read(req, resp));
  MappedMemoryRegion lba_1_and_2 =
      MappedMemoryRegion::FromCapability(resp.memory());
  uint16_t* mbr_sig = reinterpret_cast<uint16_t*>(lba_1_and_2.vaddr() + 0x1FE);
  if (*mbr_sig != 0xAA55) {
    dbgln("Invalid MBR Sig: {x}", *mbr_sig);
    return glcr::FAILED_PRECONDITION;
  }
  MbrPartition* first_partition =
      reinterpret_cast<MbrPartition*>(lba_1_and_2.vaddr() + 0x1BE);
  if (first_partition->boot_indicator != 0) {
    dbgln("Boot indicator set: {}", first_partition->boot_indicator);
    return glcr::FAILED_PRECONDITION;
  }
  if (first_partition->os_type != 0xEE) {
    dbgln("Incorrect OS type: {x}", first_partition->os_type);
    return glcr::FAILED_PRECONDITION;
  }
#if GPT_DEBUG
  dbgln("LBAs: ({x}, {x})", first_partition->starting_lba,
        first_partition->ending_lba);
#endif

  // FIXME: Don't hardcode sector size.
  ParititionHeader* header =
      reinterpret_cast<ParititionHeader*>(lba_1_and_2.vaddr() + 512);

  uint64_t num_partitions = header->num_partitions;
  uint64_t entry_size = header->parition_entry_size;
  uint64_t num_blocks = (num_partitions * entry_size) / 512;

#if GPT_DEBUG
  dbgln("signature {}", header->signature);
  dbgln("lba_partition_entries {x}", header->lba_partition_entries);
  dbgln("num_partitions: {x}", num_partitions);
  dbgln("partition_entry_size: {x}", entry_size);
  dbgln("Num blocks: {x}", num_blocks);
#endif

  req.set_device_id(0);
  req.set_lba(header->lba_partition_entries);
  req.set_size(num_blocks);
  RET_ERR(denali_->Read(req, resp));
  MappedMemoryRegion part_table =
      MappedMemoryRegion::FromCapability(resp.memory());
  for (uint64_t i = 0; i < num_partitions; i++) {
    PartitionEntry* entry = reinterpret_cast<PartitionEntry*>(
        part_table.vaddr() + (i * entry_size));
    if (entry->type_guid_low != 0 || entry->type_guid_high != 0) {
#if GPT_DEBUG
      dbgln("Entry {}", i);
      dbgln("T Guid: {x}-{x}", entry->type_guid_high, entry->type_guid_low);
      dbgln("P Guid: {x}-{x}", entry->part_guid_high, entry->part_guid_low);
      dbgln("LBA: {x}, {x}", entry->lba_start, entry->lba_end);
      dbgln("Attrs: {x}", entry->attributes);
#endif
      // For now we hardcode these values to the type that is
      // created in our setup script.
      // FIXME: Set up our own root partition type guid at some
      // point.
      if (entry->type_guid_low == kLfsDataLow &&
          entry->type_guid_high == kLfsDataHigh) {
        primary_partition_lba_ = entry->lba_start;
      }
    }
  }

  return glcr::OK;
}

File: rust/sys/yellowstone/src/gpt.rs
use denali::DenaliClient;
use mammoth::{cap::Capability, zion::ZError};

const MBR_SIG: u16 = 0xAA55;

const GPT_OS_TYPE: u8 = 0xEE;

#[repr(C, packed)]
struct MbrPartition {
    boot_indicator: u8,
    starting_chs: [u8; 3],
    os_type: u8,
    ending_chs: [u8; 3],
    starting_lba: u32,
    ending_lba: u32,
}

#[repr(C, packed)]
struct PartitionHeader {
    signature: u64,
    revision: u32,
    header_size: u32,
    crc_32: u32,
    reserved: u32,
    lba_self: u64,
    lba_mirror: u64,
    lba_min: u64,
    lba_max: u64,
    guid_low: u64,
    guid_high: u64,
    lba_partition_entries: u64,
    num_partitions: u32,
    partition_entry_size: u32,
    partition_entry_crc32: u32,
}

#[repr(C, packed)]
struct PartitionEntry {
    type_guid_low: u64,
    type_guid_high: u64,
    part_guid_low: u64,
    part_guid_high: u64,
    lba_start: u64,
    lba_end: u64,
    attrs: u64,
    partition_name: [u8; 72],
}

pub fn read_gpt(mut denali: DenaliClient) -> Result<u64, ZError> {
    let resp = denali.read(&denali::ReadRequest {
        device_id: 0,
        block: denali::DiskBlock { lba: 0, size: 2 },
    })?;

    let first_lbas = mammoth::mem::MemoryRegion::from_cap(Capability::take(resp.memory))?;

    let maybe_mbr_sig = u16::from_le_bytes(first_lbas.slice()[0x1FE..0x200].try_into().unwrap());

    if maybe_mbr_sig != MBR_SIG {
        mammoth::debug!("MBR Sig {:#x} does not match {:#x}", maybe_mbr_sig, MBR_SIG);
        return Err(ZError::FAILED_PRECONDITION);
    }

    let mbr = unsafe {
        first_lbas.slice::<u8>()[0x1BE..]
            .as_ptr()
            .cast::<MbrPartition>()
            .as_ref()
            .unwrap()
    };

    if mbr.os_type != GPT_OS_TYPE {
        let os_type = mbr.os_type;
        mammoth::debug!(
            "MBR OS Type: {:#x} does not match GPT ({:#x})",
            os_type,
            GPT_OS_TYPE
        );
    }

    let gpt_header = unsafe {
        first_lbas.slice::<u8>()[512..]
            .as_ptr()
            .cast::<PartitionHeader>()
            .as_ref()
            .unwrap()
    };

    let num_partitions = gpt_header.num_partitions;
    let partition_entry_size = gpt_header.partition_entry_size;
    let lba_partition_entries = gpt_header.lba_partition_entries;
    let num_blocks = (num_partitions * partition_entry_size).div_ceil(512) as u64;

    mammoth::debug!(
        "Reading partition table from LBA {}, {} entries of size {} equals {} blocks.",
        lba_partition_entries,
        num_partitions,
        partition_entry_size,
        num_blocks
    );

    let resp = denali.read(&denali::ReadRequest {
        device_id: 0,
        block: denali::DiskBlock {
            lba: lba_partition_entries,
            size: num_blocks,
        },
    })?;

    let partition_table = mammoth::mem::MemoryRegion::from_cap(Capability::take(resp.memory))?;

    for ind in 0..num_partitions {
        let offset = (ind * partition_entry_size) as usize;
        let limit = offset + (partition_entry_size as usize);

        let partition_entry = unsafe {
            partition_table.slice::<u8>()[offset..limit]
                .as_ptr()
                .cast::<PartitionEntry>()
                .as_ref()
                .unwrap()
        };

        if partition_entry.type_guid_low == 0 && partition_entry.type_guid_high == 0 {
            continue;
        }

        const LFS_DATA_LOW: u64 = 0x477284830fc63daf;
        const LFS_DATA_HIGH: u64 = 0xe47d47d8693d798e;
        if partition_entry.type_guid_low == LFS_DATA_LOW
            && partition_entry.type_guid_high == LFS_DATA_HIGH
        {
            return Ok(partition_entry.lba_start);
        }
    }

    Err(ZError::NOT_FOUND)
}
